x/auth/legacy/v0_39/types.go

```diff
@@ -0,0 +1,430 @@
+		Name          string         `json:"name" yaml:"name"`
+		Permissions   []string       `json:"permissions" yaml:"permissions"`
+	}
+)
+
+func NewGenesisState(params v034auth.Params, accounts v038auth.GenesisAccounts) GenesisState {
+	return GenesisState{
+		Params:   params,
+		Accounts: accounts,
+	}
+}
+
+func NewBaseAccountWithAddress(addr sdk.AccAddress) BaseAccount {
+	return BaseAccount{
+		Address: addr,
+	}
+}
+
+func NewBaseAccount(
+	address sdk.AccAddress, coins sdk.Coins, pk crypto.PubKey, accountNumber, sequence uint64,
+) *BaseAccount {
+
+	return &BaseAccount{
+		Address:       address,
+		Coins:         coins,
+		PubKey:        pk,
+		AccountNumber: accountNumber,
+		Sequence:      sequence,
+	}
+}
+
+func (acc BaseAccount) GetAddress() sdk.AccAddress {
+	return acc.Address
+}
+
+func (acc *BaseAccount) GetAccountNumber() uint64 {
+	return acc.AccountNumber
+}
+
+func (acc *BaseAccount) GetCoins() sdk.Coins {
+	return acc.Coins
+}
+
+func (acc *BaseAccount) SetCoins(coins sdk.Coins) error {
+	acc.Coins = coins
+	return nil
+}
+
+func (acc BaseAccount) Validate() error {
+	if acc.PubKey != nil && acc.Address != nil &&
+		!bytes.Equal(acc.PubKey.Address().Bytes(), acc.Address.Bytes()) {
+		return errors.New("pubkey and address pair is invalid")
+	}
+
+	return nil
+}
+
+func NewBaseVestingAccount(
+	baseAccount *BaseAccount, originalVesting, delegatedFree, delegatedVesting sdk.Coins, endTime int64,
+) *BaseVestingAccount {
+
+	return &BaseVestingAccount{
+		BaseAccount:      baseAccount,
+		OriginalVesting:  originalVesting,
+		DelegatedFree:    delegatedFree,
+		DelegatedVesting: delegatedVesting,
+		EndTime:          endTime,
+	}
+}
+
+func (bva BaseVestingAccount) MarshalJSON() ([]byte, error) {
+	alias := vestingAccountJSON{
+		Address:          bva.Address,
+		Coins:            bva.Coins,
+		PubKey:           bva.PubKey,
+		AccountNumber:    bva.AccountNumber,
+		Sequence:         bva.Sequence,
+		OriginalVesting:  bva.OriginalVesting,
+		DelegatedFree:    bva.DelegatedFree,
+		DelegatedVesting: bva.DelegatedVesting,
+		EndTime:          bva.EndTime,
+	}
+
+	return codec.Cdc.MarshalJSON(alias)
+}
+
+func (bva *BaseVestingAccount) UnmarshalJSON(bz []byte) error {
+	var alias vestingAccountJSON
+	if err := codec.Cdc.UnmarshalJSON(bz, &alias); err != nil {
+		return err
+	}
+
+	bva.BaseAccount = NewBaseAccount(alias.Address, alias.Coins, alias.PubKey, alias.AccountNumber, alias.Sequence)
+	bva.OriginalVesting = alias.OriginalVesting
+	bva.DelegatedFree = alias.DelegatedFree
+	bva.DelegatedVesting = alias.DelegatedVesting
+	bva.EndTime = alias.EndTime
+
+	return nil
+}
+
+func (bva BaseVestingAccount) GetEndTime() int64 {
+	return bva.EndTime
+}
+
+func (bva BaseVestingAccount) Validate() error {
+	if (bva.Coins.IsZero() && !bva.OriginalVesting.IsZero()) ||
+		bva.OriginalVesting.IsAnyGT(bva.Coins) {
+		return errors.New("vesting amount cannot be greater than total amount")
+	}
```

Comment:
```suggestion

```

Reply:
@clevinson @ethanfrey please review and approve this

Reply:
This change looks good to me.

Reply:
This seems straight forward, OriginalVesting and Coins have no relation.

Looking at https://github.com/cosmos/cosmos-sdk/blob/release/v0.39.1/x/auth/vesting/types/vesting_account.go#L38-L41 it seems the only invariant we could enforce are `bva.DelegatedVesting < bva.OriginalVesting`. I doubt it is worth it, but other ones don't really seem like invariants